Add missing business rules from Specify 6

melton-jason opened the pull request:
Fixes #4866

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [ ] Add automated tests
- [ ] Add relevant issue to release milestone

### Testing instructions

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->

grantfitzsimmons commented:
@CarolineDenis Should this be added to the timeline?

CarolineDenis commented:
@melton-jason 

> Should we make a migration to include these rules for instances which already have uniqueness rules created?

==> yes

combs-a commented:
I feel like on the uniqueness rule migration part, I don't like the idea that it might mess with pre-existing workflows. If it's missed in the patch notes/not relevant to users for awhile, people might be become confused when they finally run into the changed behavior.

I think that skipping creating the rules that would already be invalid/be duplicates will cause less issues. I think the migration is necessary though, so hopefully the ones that don't already have business rules now having them won't break anything 🙏

melton-jason commented:
Hey everyone! A migration has been added to this PR to create the uniqueness rules! A rule is created _only if_ there are not pre-existing duplicates for the rule. 

The PR description and testing instructions have been updated! 

Please note the first testing instruction (also provided below), which has a step that should be taken _before_ migrations have been run on a database!

> - 💡 Before creating a copy of the database for migrations, ensure that at least one of the uniqueness rules in the above "Uniqueness Rule" section will be invalid
>   - For an example, create two Disposals with the same disposalNumber, or add two AccessionAuthorizations with the same permit to an Accession, etc.

emenslin reviewed (changes requested):
- [x] Ensure that the added uniqueness rules (detailed in the above "Uniqueness Rule" section) are created when the rule is enforcable (no duplicates before running the migration)
- [x] Ensure that uniqueness rules which are not enforceable (already have duplicates) are skipped and not created

- Open a new Agent form
- [ ] Ensure agentType correctly defaults to "Person" 
- [ ] Save the new Agent, reload the page, and ensure that the Agent still has a "Person" agentType
- [x] Ensure that the previous agentType is respected


-  On the Determination form (which can be rendered as a Subview on CollectionObject), add a new Determiner
- [x] Ensure that the `Determiner -> isPrimary` is checked
- [x] Ensure that the new Determiner now has `isPrimary` checked and the old Determiner no longer has `isPrimary` checked
- On the CollectingTrip form (which can be rendered as a Subview on the CollectingEvent form), add a new Funding Agent
- [x] Ensure that the `FundingAgent -> isPrimary` is checked
- [x] Ensure that the new FundingAgent now has `isPrimary` checked and the old FundingAgent no longer has `isPrimary` checked

--- 
Everything looks good except agents do not default to person

https://github.com/user-attachments/assets/ee9cd11b-f9d0-471f-aa20-fb623407cf50

emenslin reviewed (dismissed):
- [x] Ensure that the added uniqueness rules (detailed in the above "Uniqueness Rule" section) are created when the rule is enforcable (no duplicates before running the migration)
- [x] Ensure that uniqueness rules which are not enforceable (already have duplicates) are skipped and not created

- [x] Ensure agentType correctly defaults to "Person" 
- [x] Save the new Agent, reload the page, and ensure that the Agent still has a "Person" agentType
- [x] Ensure that the previous agentType is respected
- Add a default other than `Person` to the field on the form definition
- [ ] Open a new Agent form and ensure that the field correctly defaults to the value specified on the form definition
- [x] Please also test other PickLists (agentTitle, prepType, the default "Department" picklist, etc.) and ensure that no new bugs have been introduced 🙏 



- [x] Ensure that the `Determiner -> isPrimary` is checked
- [x] Ensure that the new Determiner now has `isPrimary` checked and the old Determiner no longer has `isPrimary` checked
- [x] Ensure that the `FundingAgent -> isPrimary` is checked
- [x] Ensure that the new FundingAgent now has `isPrimary` checked and the old FundingAgent no longer has `isPrimary` checked

--- 

Cannot set a default for the agent type picklist but you can for other picklists.

Issue Branch: 

https://github.com/user-attachments/assets/59cdd451-902e-4a3e-86f7-764c21f95050


Main: 

https://github.com/user-attachments/assets/edb5d517-f3f5-4c8d-bd66-68152bb73518

melton-jason commented:
@emenslin 
Yes you're right! It looks like the business rule to set the default type is always being executed before the logic which sets the default based on the form definition. This is largely a systematic Issue with Specify currently: there currently is no concept of "precedence"--- whichever gets set first will be the value for the field. When Specify then checks to see if it should set the default value for the other value (on the form definition in this case), it sees the field already has a value and thus doesn't get set. 

In cases like these, I would always assume the default value specified on the form definition should overwrite any values set on resource creation/initialization of business rules. Would you agree? 
I'll get to work on a fix which applies the above assumption, but let me know your thoughts on it!

emenslin commented:
> In cases like these, I would always assume the default value specified on the form definition should overwrite any values set on resource creation/initialization of business rules. Would you agree?

Yes, I think that the value the user specifies in the form definition should override any of the predefined defaults. I'm thinking about how it works in other areas and in most cases when the user sets something themselves it will always reference that and not the defaults (i.e. if a user makes their own CO form definition then the form will look like that and not the default CO form).

melton-jason commented:
Hey everyone! 
I think we are going to push the setting of Person as default agentType to the next release. It may take some additional time to properly resolve the Issue, and it may be beneficial to get the other changes this branch provides into `main` and the soon-to-be `v7.11` release. 

There will a PR open to address the setting of conflicting defaults (as described in https://github.com/specify/specify7/pull/4892#issuecomment-2904421685) soon! 

For now, can this be tested again without that functionality? I have removed it from this branch and updated the testing instrunctions.

grantfitzsimmons commented:
@melton-jason not so sure [`213edbf` (#4892)](https://github.com/specify/specify7/pull/4892/commits/213edbf72be9ed6f7e847c91cf8eb78b32952307) is working anymore, create a new agent on sp7demofish and see: https://sp7demofish.specifycloud.org/specify/view/agent/new/
